File: go/Assembler/config/config_util.go
package config

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

type Scenario struct {
	AppSite string
	AppFile string
	AppView string
}

func NewScenario(appSite, appFile, appView string) Scenario {
	return Scenario{
		AppSite: appSite,
		AppFile: appFile,
		AppView: appView,
	}
}

func (s *Scenario) ToString() string {
	return fmt.Sprintf("%s:%s:%s", s.AppSite, s.AppFile, s.AppView)
}

type configCache struct {
	wwwrootPath string
	appSites    map[string]bool
	scenarios   []Scenario
}

var (
	cache      *configCache
	cacheMutex sync.Mutex
)

type ConfigUtil struct{}

const DefaultAppFile = "Index"

// extractAppSitesFromScenarios extracts unique AppSites from scenarios
func extractAppSitesFromScenarios(scenarios []Scenario) map[string]bool {
	appSitesMap := make(map[string]bool)
	for _, scenario := range scenarios {
		if scenario.AppSite != "" {
			appSitesMap[strings.ToLower(scenario.AppSite)] = true
		}
	}

	fmt.Printf("[ConfigUtil] Extracted %d AppSites from folder scan\n", len(appSitesMap))
	return appSitesMap
}

// loadScenariosInternal discovers scenarios by scanning AppSites folder structure
func loadScenariosInternal(wwwrootPath string) ([]Scenario, error) {
	appSitesPath := filepath.Join(wwwrootPath, "AppSites")

	if _, err := os.Stat(appSitesPath); os.IsNotExist(err) {
		return nil, fmt.Errorf("AppSites directory not found: %s", appSitesPath)
	}

	scenarios := []Scenario{}

	// Get all directories in AppSites folder
	entries, err := os.ReadDir(appSitesPath)
	if err != nil {
		return nil, fmt.Errorf("Failed to read AppSites directory: %v", err)
	}

	var appSiteDirs []string
	for _, entry := range entries {
		if entry.IsDir() {
			appSiteDirs = append(appSiteDirs, entry.Name())
		}
	}
	sort.Strings(appSiteDirs)

	for _, appSite := range appSiteDirs {
		// Get all HTML files in the appSite directory (top level only)
		appSiteDir := filepath.Join(appSitesPath, appSite)

		files, err := os.ReadDir(appSiteDir)
		if err != nil {
			continue
		}

		var htmlFiles []string
		for _, file := range files {
			if file.IsDir() || !strings.HasSuffix(file.Name(), ".html") {
				continue
			}
			htmlFiles = append(htmlFiles, file.Name())
		}

		// If no HTML files found, use DefaultAppFile
		if len(htmlFiles) == 0 {
			htmlFiles = []string{DefaultAppFile}
		}

		for _, fileName := range htmlFiles {
			var appFile string
			if fileName == DefaultAppFile {
				appFile = DefaultAppFile
			} else {
				appFile = strings.TrimSuffix(fileName, ".html")
			}

			// Check for Views folder
			viewsPath := filepath.Join(appSiteDir, "Views")
			var viewDirs []string

			if viewsInfo, err := os.Stat(viewsPath); err == nil && viewsInfo.IsDir() {
				// Get all subdirectories in Views folder
				if viewEntries, err := os.ReadDir(viewsPath); err == nil {
					for _, viewEntry := range viewEntries {
						if viewEntry.IsDir() {
							viewDirs = append(viewDirs, viewEntry.Name())
						}
					}
				}
			}

			// Only add empty AppView scenario if no specific Views exist
			if len(viewDirs) == 0 {
				scenarios = append(scenarios, NewScenario(appSite, appFile, ""))
			} else {
				// Add specific view scenarios
				for _, viewDir := range viewDirs {
					scenarios = append(scenarios, NewScenario(appSite, appFile, viewDir))
				}
			}
		}
	}

	if len(scenarios) == 0 {
		return nil, fmt.Errorf("No scenarios found in AppSites folder")
	}

	fmt.Printf("[ConfigUtil] Loaded %d scenarios from AppSites folder\n", len(scenarios))

	return scenarios, nil
}

// Load loads AppSites from wwwroot path and caches them. Call this during startup.
func Load(wwwrootPath string) error {
	cacheMutex.Lock()
	defer cacheMutex.Unlock()

	scenarios, err := loadScenariosInternal(wwwrootPath)
	if err != nil {
		return err
	}

	appSites := extractAppSitesFromScenarios(scenarios)

	cache = &configCache{
		wwwrootPath: wwwrootPath,
		scenarios:   scenarios,
		appSites:    appSites,
	}

	return nil
}

// Reload reloads AppSites and Scenarios from the stored wwwroot path. Throws if not loaded.
func Reload() error {
	cacheMutex.Lock()
	defer cacheMutex.Unlock()

	if cache == nil {
		return fmt.Errorf("ConfigUtil not loaded. Call Load(wwwrootPath) first.")
	}

	scenarios, err := loadScenariosInternal(cache.wwwrootPath)
	if err != nil {
		return err
	}

	appSites := extractAppSitesFromScenarios(scenarios)

	cache.scenarios = scenarios
	cache.appSites = appSites

	return nil
}

// GetAppSites gets the cached AppSites. Throws if not loaded.
func GetAppSites() (map[string]bool, error) {
	cacheMutex.Lock()
	defer cacheMutex.Unlock()

	if cache == nil {
		return nil, fmt.Errorf("AppSitesConfig not loaded. Call Load(wwwrootPath) first.")
	}

	return cache.appSites, nil
}

// GetScenarios gets the cached Scenarios. Throws if not loaded.
func GetScenarios() ([]Scenario, error) {
	cacheMutex.Lock()
	defer cacheMutex.Unlock()

	if cache == nil {
		return nil, fmt.Errorf("AppSitesConfig not loaded. Call Load(wwwrootPath) first.")
	}

	return cache.scenarios, nil
}

// FilterByAppSite filters scenarios by appSite
func FilterByAppSite(scenarios []Scenario, appSiteFilter string) []Scenario {
	if appSiteFilter == "" {
		return scenarios
	}

	filtered := []Scenario{}
	for _, s := range scenarios {
		if strings.EqualFold(s.AppSite, appSiteFilter) {
			filtered = append(filtered, s)
		}
	}
	return filtered
}

File: php/Assembler/src/Config/ConfigUtil.php
<?php

namespace Assembler\Config;

class Scenario
{
    public string $appSite;
    public string $appFile;
    public string $appView;

    public function __construct(string $appSite, string $appFile, string $appView)
    {
        $this->appSite = $appSite;
        $this->appFile = $appFile;
        $this->appView = $appView;
    }

    public function __toString(): string
    {
        return "{$this->appSite}:{$this->appFile}:{$this->appView}";
    }
}

class ConfigUtil
{
    public static string $defaultAppFile = 'Index';

    private static ?array $cachedAppSites = null;
    private static ?array $cachedScenarios = null;
    private static ?string $wwwrootPath = null;

    /**
     * Extracts unique AppSites from scenarios
     */
    private static function extractAppSitesFromScenarios(array $scenarios): array
    {
        $appSitesSet = [];
        foreach ($scenarios as $scenario) {
            if (!empty($scenario->appSite)) {
                $appSitesSet[strtolower($scenario->appSite)] = true;
            }
        }

        echo sprintf("[ConfigUtil] Extracted %d AppSites from folder scan\n", count($appSitesSet));

        return array_keys($appSitesSet);
    }

    /**
     * Discovers scenarios by scanning AppSites folder structure
     */
    private static function loadScenariosInternal(string $wwwrootPath): array
    {
        $appSitesPath = $wwwrootPath . DIRECTORY_SEPARATOR . 'AppSites';

        if (!is_dir($appSitesPath)) {
            throw new \Exception("AppSites directory not found: $appSitesPath");
        }

        $scenarios = [];

        // Get all directories in AppSites folder
        $entries = scandir($appSitesPath);
        $appSiteDirs = [];

        foreach ($entries as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $fullPath = $appSitesPath . DIRECTORY_SEPARATOR . $entry;
            if (is_dir($fullPath)) {
                $appSiteDirs[] = $entry;
            }
        }

        sort($appSiteDirs);

        foreach ($appSiteDirs as $appSite) {
            // Get all HTML files in the appSite directory (top level only)
            $appSiteDir = $appSitesPath . DIRECTORY_SEPARATOR . $appSite;
            $files = scandir($appSiteDir);

            $htmlFiles = [];
            foreach ($files as $file) {
                $filePath = $appSiteDir . DIRECTORY_SEPARATOR . $file;
                if (is_file($filePath) && str_ends_with($file, '.html')) {
                    $htmlFiles[] = $file;
                }
            }

            // If no HTML files found, use DEFAULT_APP_FILE
            if (empty($htmlFiles)) {
                $htmlFiles = [self::$defaultAppFile];
            }

            foreach ($htmlFiles as $htmlFile) {
                $appFile = $htmlFile === self::$defaultAppFile ? self::$defaultAppFile : pathinfo($htmlFile, PATHINFO_FILENAME);

                // Check for Views folder
                $viewsPath = $appSiteDir . DIRECTORY_SEPARATOR . 'Views';
                $viewDirs = [];

                if (is_dir($viewsPath)) {
                    // Get all subdirectories in Views folder
                    $viewEntries = scandir($viewsPath);
                    foreach ($viewEntries as $viewEntry) {
                        if ($viewEntry === '.' || $viewEntry === '..') {
                            continue;
                        }
                        $viewPath = $viewsPath . DIRECTORY_SEPARATOR . $viewEntry;
                        if (is_dir($viewPath)) {
                            $viewDirs[] = $viewEntry;
                        }
                    }
                }

                // Only add empty AppView scenario if no specific Views exist
                if (count($viewDirs) === 0) {
                    $scenarios[] = new Scenario($appSite, $appFile, '');
                } else {
                    // Add specific view scenarios
                    foreach ($viewDirs as $viewDir) {
                        $scenarios[] = new Scenario($appSite, $appFile, $viewDir);
                    }
                }
            }
        }

        if (count($scenarios) === 0) {
            throw new \Exception('No scenarios found in AppSites folder');
        }

        echo sprintf("[ConfigUtil] Loaded %d scenarios from AppSites folder\n", count($scenarios));

        return $scenarios;
    }

    /**
     * Loads AppSites from wwwroot path and caches them. Call this during startup.
     */
    public static function load(string $wwwrootPath): void
    {
        self::$wwwrootPath = $wwwrootPath;
        self::$cachedScenarios = self::loadScenariosInternal($wwwrootPath);
        self::$cachedAppSites = self::extractAppSitesFromScenarios(self::$cachedScenarios);
    }

    /**
     * Reloads AppSites and Scenarios from the stored wwwroot path. Throws if not loaded.
     */
    public static function reload(): void
    {
        if (self::$wwwrootPath === null) {
            throw new \Exception('ConfigUtil not loaded. Call load(wwwrootPath) first.');
        }

        self::$cachedScenarios = self::loadScenariosInternal(self::$wwwrootPath);
        self::$cachedAppSites = self::extractAppSitesFromScenarios(self::$cachedScenarios);
    }

    /**
     * Gets the cached AppSites. Throws if not loaded.
     */
    public static function getAppSites(): array
    {
        if (self::$cachedAppSites === null) {
            throw new \Exception('AppSitesConfig not loaded. Call load(wwwrootPath) first.');
        }

        return self::$cachedAppSites;
    }

    /**
     * Gets the cached Scenarios. Throws if not loaded.
     */
    public static function getScenarios(): array
    {
        if (self::$cachedScenarios === null) {
            throw new \Exception('AppSitesConfig not loaded. Call load(wwwrootPath) first.');
        }

        return self::$cachedScenarios;
    }

    /**
     * Filters scenarios by appSite
     */
    public static function filterByAppSite(array $scenarios, ?string $appSiteFilter): array
    {
        if (empty($appSiteFilter)) {
            return $scenarios;
        }

        return array_values(array_filter(
            $scenarios,
            fn($s) => strcasecmp($s->appSite, $appSiteFilter) === 0
        ));
    }
}
